### ISSUE TITLE
Variables defined in while(true) loop are treated as possibly undefined.

### ISSUE THREAD

**Issue by TysonAndre** [vimeo/psalm#381]
The real example had a conditional break instead of an unconditional break.

The way these are analyzed to check if a variable might be undefined seems to be analyzed differently from a `do{} while`, which doesn't warn about being possibly undefined.

If a variable is assigned (in all branches) before any break statements and the loop condition is unconditionally true, I'd expect psalm to infer it was guaranteed to be defined (Ignoring infinite loops)

This is also infrequent. A workaround is to define the variable above the loop.

```php
<?php
function guaranteedloop() : void {
    while (true) {
        $x = 'value';
        break;
    }
    echo "$x\n";  // PossiblyUndefinedVariable - Possibly undefined variable $x, first seen on line 4
    for (;;) {
        $y = 'value';
        break;
    }
    echo "$y\n";  // PossiblyUndefinedVariable - Possibly undefined variable $y...
}
```

**Comment by muglug** [comment]
Yeah, I can definitely create a special case. If you rewrite with `do {} while (true)` Psalm will understand as well

**Comment by weirdan** [comment]
> If you rewrite with `do {} while (true)`

That should actually work with `do {} while (anything);` as well (including `do {} while (false);`.

**Comment by muglug** [comment]
Oh, absolutely. And it does. But the former actually works the same way as `while (true)`
